pxr/base/tf/unicodeUtils.h

```diff
@@ -0,0 +1,625 @@
+            return false;
+        }
+
+        // remaining characters
+        utf8Iterator++;
+        for (; utf8Iterator != end; utf8Iterator++)
+        {
+            codePoint = *utf8Iterator;
+            bool found = (xidStartClass.find(codePoint) != xidStartClass.end()) || (xidContinueClass.find(codePoint) != xidContinueClass.end());
+            if (!found)
+            {
+                // not in the singular set, need to check the range sets
+                for (size_t i = 0; i < xidStartRangeClass.size() && !found; i++)
+                {
+                    found = (codePoint >= xidStartRangeClass[i].first && codePoint <= xidStartRangeClass[i].second);
+                }
+
+                if (!found)
+                {
+                    for (size_t i = 0; i < xidContinueRangeClass.size() && !found; i++)
+                    {
+                        found = (codePoint >= xidContinueRangeClass[i].first && codePoint <= xidContinueRangeClass[i].second);
+                    }
+                }
+            }
+
+            if (!found)
+            {
+                // it's not in either the XID_Start class or the XID_Continue class so it's invalid
+                return false;
+            }
+        }
+
+        return (utf8Iterator == end);
+    }
+
+    /// Determines whether the UTF-8 encoded substring in \a identifier starting at position \a sequenceStart and ending at position \a end is a
+    /// valid Unicode prim name. Prim names are more permissive than identifier names in that all characters can be part of the XID_Continue
+    /// character class, meaning that prim names could start with digits (or be all digits), etc.
+    ///
+    /// The XID_Start class of characters are derived from the Unicode General_Category of uppercase letters, lowercase letters, titlecase letters,
+    /// modifier letters, other letters, letters numbers, plus Other_ID_Start, minus Pattern_Syntax and Pattern_White_Space code points.
+    /// That is, the character must have a category of Lu | Ll | Lt | Lm | Lo | Nl | '_'
+    ///
+    /// The XID_Continue class of characters include those in XID_Start plus characters having the Unicode General Category of nonspacing marks,
+    /// spacing combining marks, decimal number, and connector punctuation.
+    /// That is, the character must have a category of XID_Start | Nd | Mn | Mc | Pc
+    ///
+    /// UTF-8 characters are variable encoded, so \a sequenceStart defines the first byte in the UTF-8 character sequence. This method can be used for entire
+    /// strings by passing identifier.begin() and identifier.end(), but also on sub sequences defined by the given iterator ranges to avoid copying
+    /// the subsequence to a temporary string for evaluation.
+    ///
+    inline TF_API bool IsValidUTF8PrimName(const std::string& identifier, const std::string::const_iterator& sequenceStart, const std::string::const_iterator& end)
```

Comment:
Alternatively, since the rules for prim names are just valid sequences of XID_Continue / XID_Start, you could probably come up with a more generic  utility name that doesn't explicitly reference prims.

Reply:
I'm just going to remove the "Prim" part of this to turn it into `IsValidUTF8Name`